### PR TITLE
Align with loss func of Megatron

### PR DESCRIPTION
**Pull request by SaoXiaoPiZhuPi**
Change the loss function in examples/llama/train.py, so that the implementation of loss function is the same with Megatron, which can be slightly faster.

### CODE REVIEW

**Comment by J-shang** [comment]
@SaoXiaoPiZhuPi, thanks for your contribution

1. could you point out different of
```
loss_fct = torch.nn.CrossEntropyLoss(ignore_index=IGNORE_IDX)
loss = loss_fct(logits, labels)
```
and
```
loss = torch.nn.functional.cross_entropy(logits, labels, reduction='mean', ignore_index=IGNORE_IDX)
```
Although there is no essential difference between these two types of writing for nnscaler, because nnscaler is a function level compiler, the second way is prefered.

2. could you point out the scenario the speedup would be significant between these? I test the several writing types, seems there is no significant speed different:
```
import torch

def test_time(func):
    logits = torch.rand(8196, 128256, requires_grad=True, device="cuda")
    labels = torch.randint(0, 128256, [8196], device="cuda")
    torch.cuda.synchronize()

    start_event = torch.cuda.Event(enable_timing=True)
    end_event = torch.cuda.Event(enable_timing=True)

    start_event.record()
    func(logits, labels)
    end_event.record()
    torch.cuda.synchronize()
    elapsed_time_ms = start_event.elapsed_time(end_event)
    return elapsed_time_ms


def combine_fw(logits, labels):
    for i in range(100):
        normalized_logits = torch.nn.functional.log_softmax(logits, dim=-1, dtype=torch.float32)
        loss = torch.nn.functional.nll_loss(normalized_logits, labels, reduction='sum', ignore_index=-100)

def direct_fw(logits, labels):
    for i in range(100):
        loss = torch.nn.functional.cross_entropy(logits, labels, reduction='sum', ignore_index=-100)

def direct_fw_mean(logits, labels):
    for i in range(100):
        loss = torch.nn.functional.cross_entropy(logits, labels, reduction='mean', ignore_index=-100)

def direct_mod_fw_mean(logits, labels):
    for i in range(100):
        loss_func = torch.nn.CrossEntropyLoss(ignore_index=-100)
        loss = loss_func(logits, labels)

def combine_fw_bw(logits, labels):
    for i in range(100):
        normalized_logits = torch.nn.functional.log_softmax(logits, dim=-1, dtype=torch.float32)
        loss = torch.nn.functional.nll_loss(normalized_logits, labels, reduction='sum', ignore_index=-100)
        loss.backward()

def direct_fw_bw(logits, labels):
    for i in range(100):
        loss = torch.nn.functional.cross_entropy(logits, labels, reduction='sum', ignore_index=-100)
        loss.backward()

def direct_fw_bw_mean(logits, labels):
    for i in range(100):
        loss = torch.nn.functional.cross_entropy(logits, labels, reduction='mean', ignore_index=-100)
        loss.backward()

def direct_mod_fw_bw_mean(logits, labels):
    for i in range(100):
        loss_func = torch.nn.CrossEntropyLoss(ignore_index=-100)
        loss = loss_func(logits, labels)
        loss.backward()

print(f"combine_fw took {test_time(combine_fw):.3f} ms on CUDA.")
print(f"direct_fw took {test_time(direct_fw):.3f} ms on CUDA.")
print(f"direct_fw_mean took {test_time(direct_fw_mean):.3f} ms on CUDA.")
print(f"direct_mod_fw_mean took {test_time(direct_mod_fw_mean):.3f} ms on CUDA.")
print(f"combine_fw_bw took {test_time(combine_fw_bw):.3f} ms on CUDA.")
print(f"direct_fw_bw took {test_time(direct_fw_bw):.3f} ms on CUDA.")
print(f"direct_fw_bw_mean took {test_time(direct_fw_bw_mean):.3f} ms on CUDA.")
print(f"direct_mod_fw_bw_mean took {test_time(direct_mod_fw_bw_mean):.3f} ms on CUDA.")
```
```
combine_fw took 1094.301 ms on CUDA.
direct_fw took 1066.114 ms on CUDA.
direct_fw_mean took 1066.264 ms on CUDA.
direct_mod_fw_mean took 1066.279 ms on CUDA.
combine_fw_bw took 3105.789 ms on CUDA.
direct_fw_bw took 3104.792 ms on CUDA.
direct_fw_bw_mean took 3104.748 ms on CUDA.
direct_mod_fw_bw_mean took 3105.042 ms on CUDA.
```

**Comment by SaoXiaoPiZhuPi** [comment]
Sorry for taking up your time. The reason I previously thought about aligning the loss function API is that I tried to compare the cases where tp == 1 and pp == 1 to align the configurations of Megatron and nnScaler. When I used *torch.profiler8 to trace the model, I found that the overhead of the cunn_SoftMaxForward operator for the two methods was inconsistent. The original nnScaler method took 2455us, which is almost twice that of the Megatron method (1266us). However, after the alignment, the overhead of cunn_SoftMaxForward for nnScaler and Megatron became consistent. 

After receiving your reply, I switched to a new configuration and retested, but the results showed that the original nnScaler method was actually faster than the Megatron method (the performance gain should come from the benefits of parallelism mentioned by @zyeric).

Thank you again for your clarification. I think we can close this PR.

**Comment by zyeric** [comment]
Thanks for your profiling and information. I will close this PR. Let's make nnScaler better together :)
